File: go/internal/harness/params.go
package harness

import (
	"fmt"
	"strconv"
	"strings"
)

// Params holds the --param key=value pairs passed by the orchestrator.
// Sizes arrive already converted to plain integers (bytes, elements, ...).
type Params map[string]string

// Str returns a string parameter or def when absent.
func (p Params) Str(key, def string) string {
	if v, ok := p[key]; ok {
		return v
	}
	return def
}

// Int returns an integer parameter or def when absent. It panics on a
// malformed value: a bad parameter is a harness bug, never a data point.
func (p Params) Int(key string, def int64) int64 {
	v, ok := p[key]
	if !ok {
		return def
	}
	n, err := strconv.ParseInt(strings.ReplaceAll(v, "_", ""), 10, 64)
	if err != nil {
		panic(fmt.Sprintf("param %s=%q: %v", key, v, err))
	}
	return n
}

// Float returns a float parameter or def when absent.
func (p Params) Float(key string, def float64) float64 {
	v, ok := p[key]
	if !ok {
		return def
	}
	f, err := strconv.ParseFloat(v, 64)
	if err != nil {
		panic(fmt.Sprintf("param %s=%q: %v", key, v, err))
	}
	return f
}

// Bool returns a boolean parameter or def when absent.
func (p Params) Bool(key string, def bool) bool {
	v, ok := p[key]
	if !ok {
		return def
	}
	b, err := strconv.ParseBool(v)
	if err != nil {
		panic(fmt.Sprintf("param %s=%q: %v", key, v, err))
	}
	return b
}

// MustStr returns a required string parameter.
func (p Params) MustStr(key string) string {
	v, ok := p[key]
	if !ok || v == "" {
		panic(fmt.Sprintf("missing required param %q", key))
	}
	return v
}

// paramFlag implements flag.Value for repeatable --param key=value flags.
type paramFlag Params

func (f paramFlag) String() string { return "" }

func (f paramFlag) Set(s string) error {
	k, v, ok := strings.Cut(s, "=")
	if !ok || k == "" {
		return fmt.Errorf("want key=value, got %q", s)
	}
	f[k] = v
	return nil
}

File: rust/common/src/params.rs
//! `--param key=value` pairs, mirroring go/internal/harness/params.go.

use std::collections::BTreeMap;

#[derive(Clone, Debug, Default)]
pub struct Params(pub BTreeMap<String, String>);

impl Params {
    pub fn str_or(&self, key: &str, def: &str) -> String {
        self.0.get(key).cloned().unwrap_or_else(|| def.to_string())
    }

    /// A required string parameter.
    pub fn must_str(&self, key: &str) -> String {
        match self.0.get(key) {
            Some(v) if !v.is_empty() => v.clone(),
            _ => panic!("missing required param {key:?}"),
        }
    }

    /// An integer parameter (underscores allowed). Panics when malformed: a
    /// bad parameter is a harness bug, never a data point.
    pub fn int(&self, key: &str, def: i64) -> i64 {
        match self.0.get(key) {
            None => def,
            Some(v) => v.replace('_', "").parse().unwrap_or_else(|e| panic!("param {key}={v:?}: {e}")),
        }
    }

    pub fn float(&self, key: &str, def: f64) -> f64 {
        match self.0.get(key) {
            None => def,
            Some(v) => v.parse().unwrap_or_else(|e| panic!("param {key}={v:?}: {e}")),
        }
    }

    pub fn bool(&self, key: &str, def: bool) -> bool {
        match self.0.get(key).map(|s| s.as_str()) {
            None => def,
            Some("1" | "t" | "T" | "true" | "TRUE" | "True") => true,
            Some("0" | "f" | "F" | "false" | "FALSE" | "False") => false,
            Some(v) => panic!("param {key}={v:?}: invalid bool"),
        }
    }
}
